File: simulator/keyManager.js
const { generateKey } = require('./cipher');

let currentKey = generateKey();
let rotationInterval = null;

function getCurrentKey() {
  return currentKey;
}

function rotateKey() {
  const oldKey = currentKey;
  currentKey = generateKey();
  console.log(`[KeyManager] Clave rotada: ${new Date().toISOString()}`);
  return { oldKey, newKey: currentKey };
}

function startRotation(intervalMinutes = 5) {
  if (rotationInterval) clearInterval(rotationInterval);
  rotationInterval = setInterval(() => {
    rotateKey();
  }, intervalMinutes * 60 * 1000);
  console.log(`[KeyManager] Rotación automática cada ${intervalMinutes} minutos`);
}

function stopRotation() {
  if (rotationInterval) {
    clearInterval(rotationInterval);
    rotationInterval = null;
    console.log('[KeyManager] Rotación detenida');
  }
}

module.exports = { getCurrentKey, rotateKey, startRotation, stopRotation };

File: firmware/esp32_security_p2p/key_manager.cpp
// Puerto de backend/src/crypto/keyManager.js al firmware ESP32.
// Node usa setInterval(); en un microcontrolador no hay que bloquear el loop(),
// así que aquí se usa un patrón basado en millis() no bloqueante:
// keyManagerUpdate() debe llamarse en cada vuelta del loop() principal.

#include "key_manager.h"
#include "crypto.h"

static String currentKey;
static bool rotationActive = false;
static unsigned long intervalMs = 5UL * 60UL * 1000UL; // 5 minutos por defecto
static unsigned long lastRotation = 0;

void keyManagerInit() {
  currentKey = generateKey();
  lastRotation = millis();
  Serial.println("[KeyManager] Clave inicial generada");
}

String getCurrentKey() {
  return currentKey;
}

String rotateKey() {
  currentKey = generateKey();
  lastRotation = millis();
  Serial.print("[KeyManager] Clave rotada en t=");
  Serial.println(lastRotation);
  return currentKey;
}

void startRotation(unsigned long intervalMinutes) {
  intervalMs = intervalMinutes * 60UL * 1000UL;
  rotationActive = true;
  lastRotation = millis();
  Serial.print("[KeyManager] Rotación automática cada ");
  Serial.print(intervalMinutes);
  Serial.println(" minutos");
}

void stopRotation() {
  rotationActive = false;
  Serial.println("[KeyManager] Rotación detenida");
}

void keyManagerUpdate() {
  if (!rotationActive) return;
  if (millis() - lastRotation >= intervalMs) {
    rotateKey();
  }
}